### PR TITLE
PUD-1319  API: Support last known addresses

### PR DESCRIPTION
**Pull request by AndyNewton**
PUD-1319  API: Support saving and return of offender last known addresses versus Recalls
To follow: PACT uplift to test LastKnownAddresses on Recall response.

### CODE REVIEW

File: src/main/resources/db/migration/V73__Add_last_known_addresses_table.sql [hunk 1]
```diff
@@ -0,0 +1,16 @@
+CREATE TABLE last_known_address (
+          id UUID PRIMARY KEY,
+          recall_id UUID NOT NULL REFERENCES recall (id),
+          line1 TEXT NOT NULL,
+          line2 TEXT NULL,
+          town TEXT NOT NULL,
+          postcode TEXT NULL,
+          source TEXT NOT NULL,
```
Comment (AndyNewton) [inline src/main/resources/db/migration/V73__Add_last_known_addresses_table.sql:8]: Just looking to tighten this to:
          source TEXT NOT NULL CHECK (source IN (('MANUAL','LOOKUP'))),

File: src/main/kotlin/uk/gov/justice/digital/hmpps/managerecallsapi/db/LastKnownAddressRepository.kt [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+import org.springframework.stereotype.Component
+import org.springframework.stereotype.Repository
+import java.util.UUID
+
+@Repository("jpaLastKnownAddressRepository")
+interface JpaLastKnownAddressRepository : JpaRepository<LastKnownAddress, UUID>
+
+@NoRepositoryBean
+interface ExtendedLastKnownAddressRepository : JpaLastKnownAddressRepository
+
```
Comment (philsegal) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/managerecallsapi/db/LastKnownAddressRepository.kt:16]: Never sure if we use these anymore. Recall Mike adding them for some tests 🤷‍♂️

File: src/main/resources/db/migration/V73__Add_last_known_addresses_table.sql [hunk 3]
```diff
@@ -0,0 +1,16 @@
...
+          index INTEGER NOT NULL,
+          created_by_user_id UUID references user_details(id),
+          created_date_time TIMESTAMPTZ NOT NULL,
+          UNIQUE (recall_id, index),
+          CONSTRAINT index_positive_integer CHECK (index >= 1)
+);
+
+-- DROP TABLE last_known_address;
```
Comment (philsegal) [inline src/main/resources/db/migration/V73__Add_last_known_addresses_table.sql:16]: Could be deleted, but no harm

File: src/main/resources/db/migration/V73__Add_last_known_addresses_table.sql [hunk 4]
```diff
@@ -0,0 +1,16 @@
...
+          id UUID PRIMARY KEY,
+          recall_id UUID NOT NULL REFERENCES recall (id),
+          line1 TEXT NOT NULL,
+          line2 TEXT NULL,
+          town TEXT NOT NULL,
+          postcode TEXT NULL,
+          source TEXT NOT NULL,
+          index INTEGER NOT NULL,
+          created_by_user_id UUID references user_details(id),
+          created_date_time TIMESTAMPTZ NOT NULL,
```
Comment (AndyNewton) [inline src/main/resources/db/migration/V73__Add_last_known_addresses_table.sql:11]: Just looking to tighten this to:
          created_by_user_id UUID NOT NULL references user_details(id),

File: src/main/resources/db/migration/V73__Add_last_known_addresses_table.sql (hunk 3 above)
Comment (AndyNewton) [reply src/main/resources/db/migration/V73__Add_last_known_addresses_table.sql:16]: Doing

File: src/test/kotlin/uk/gov/justice/digital/hmpps/managerecallsapi/component/LastKnownAddressComponentTest.kt [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+
+    val lastKnownAddresses = updatedRecall.lastKnownAddresses
+    assertThat(lastKnownAddresses.size, equalTo(2))
+
+    assertThat(lastKnownAddresses[0].index, !equalTo(lastKnownAddresses[1].index))
+    // There is no contract for the order of LastKnownAddresses but the index must be one-based incrementing positive integers, hence 1, 2, etc.
+    val lastKnownAddressIndex1 = lastKnownAddresses[lastKnownAddresses.indexOfFirst { it.index == 1 }]
+    val lastKnownAddressIndex2 = lastKnownAddresses[lastKnownAddresses.indexOfFirst { it.index == 2 }]
+    assertThat(lastKnownAddressIndex1.lastKnownAddressId.toString(), equalTo(lastKnownAddressId1.toString()))
+    assertThat(lastKnownAddressIndex2.lastKnownAddressId.toString(), equalTo(lastKnownAddressId2.toString()))
```
Comment (philsegal) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/managerecallsapi/component/LastKnownAddressComponentTest.kt:90]: Odd that you need to toString() the UUID

File: src/test/kotlin/uk/gov/justice/digital/hmpps/managerecallsapi/integration/db/IntegrationTestBase.kt [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+        FirstName("Test"),
+        LastName("User"),
+        "",
+        Email("test@user.com"),
+        PhoneNumber("09876543210"),
+        CaseworkerBand.FOUR_PLUS,
+        OffsetDateTime.now()
+      )
+    )
+  }
```
Comment (philsegal) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/managerecallsapi/integration/db/IntegrationTestBase.kt:66]: 👍

**Review by philsegal — APPROVED** [review]

File: src/main/resources/db/migration/V73__Add_last_known_addresses_table.sql (hunk 4 above)
Comment (AndyNewton) [reply src/main/resources/db/migration/V73__Add_last_known_addresses_table.sql:11]: Added

File: src/test/kotlin/uk/gov/justice/digital/hmpps/managerecallsapi/component/LastKnownAddressComponentTest.kt (hunk 5 above)
Comment (AndyNewton) [reply src/test/kotlin/uk/gov/justice/digital/hmpps/managerecallsapi/component/LastKnownAddressComponentTest.kt]: Fixed

File: src/main/resources/db/migration/V73__Add_last_known_addresses_table.sql (hunk 1 above)
Comment (AndyNewton) [reply src/main/resources/db/migration/V73__Add_last_known_addresses_table.sql:8]: Added a named constraint for this.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/managerecallsapi/db/LastKnownAddressRepository.kt (hunk 2 above)
Comment (AndyNewton) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/managerecallsapi/db/LastKnownAddressRepository.kt:16]: 🤷‍♂️

**Review by philsegal — APPROVED** [review]

**Pull request merged**